rust/src/lib.rs

```diff
@@ -300,92 +354,92 @@ impl Vault {
-            .ok_or(VaultError::BucketNameMissingError)?,
-        key_arn: parse_output_value_from_key("kmsKeyArn", stack_output),
-        // deployed_version: parse_output_value_from_key("vaultStackVersion", &stack_output),
-    })
-}
+    /// Get CloudFormation parameters based on config and stack name
+    async fn get_cloudformation_params(
+        config: &SdkConfig,
+        stack: &str,
+    ) -> Result<CloudFormationParams, VaultError> {
+        let describe_stack_output = CloudFormationClient::new(config)
+            .describe_stacks()
+            .stack_name(stack)
+            .send()
+            .await?;
 
-impl fmt::Display for Vault {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "region: {}\n{}", self.region, self.cloudformation_params)
+        let stack_output = describe_stack_output
+            .stacks()
+            .iter()
+            .next()
+            .map(|stack| stack.outputs())
+            .ok_or(VaultError::StackOutputsMissingError)?;
+
+        let bucket_name = Self::parse_output_value_from_key("vaultBucketName", stack_output)
+            .ok_or(VaultError::BucketNameMissingError)?;
+
+        let key_arn = Self::parse_output_value_from_key("kmsKeyArn", stack_output);
+
+        Ok(CloudFormationParams::new(bucket_name, key_arn))
     }
-}
 
-fn get_region_provider(region_opt: Option<&str>) -> RegionProviderChain {
-    RegionProviderChain::first_try(region_opt.map(|r| Region::new(r.to_owned())))
-        .or_default_provider()
-}
+    fn get_region_provider(region_opt: Option<&str>) -> RegionProviderChain {
+        RegionProviderChain::first_try(region_opt.map(|r| Region::new(r.to_owned())))
+            .or_default_provider()
+    }
 
-fn parse_output_value_from_key(key: &str, out: &[Output]) -> Option<String> {
-    out.iter()
-        .find(|output| output.output_key() == Some(key))
-        .map(|output| output.output_value().unwrap_or_default().to_owned())
-}
+    fn parse_output_value_from_key(key: &str, out: &[Output]) -> Option<String> {
+        out.iter()
+            .find(|output| output.output_key() == Some(key))
+            .map(|output| output.output_value().unwrap_or_default().to_owned())
+    }
 
-/// Helper function to get all three key names for S3 data
-fn get_s3_data_keys(name: &str) -> [String; 3] {
-    [
-        format!("{name}.aesgcm.encrypted"),
-        format!("{name}.key"),
-        format!("{name}.meta"),
-    ]
-}
+    fn create_random_nonce() -> [u8; 12] {
+        let mut nonce: [u8; 12] = [0; 12];
+        let mut rng = rand::thread_rng();
+        rng.fill(nonce.as_mut_slice());
+        nonce
+    }
 
-/// Return possible env variable value as Option
-fn get_env_variable(name: &str) -> Option<String> {
-    env::var(name).ok()
+    /// Return possible env variable value as Option
+    fn get_env_variable(name: &str) -> Option<String> {
+        env::var(name).ok()
+    }
```

Comment:
I'm not sure why we even have this function, as `env::var(name).ok()` should suffice on its own. Nevertheless, should not be in `impl Vault`

Reply:
Why not, it is only used there and nowhere else. 

ChatGPT:

> Associated functions help in organizing code through namespacing. By associating functions with the types they operate on or are related to, Rust allows for cleaner code organization and easier navigation. It becomes immediately clear which type a function is meant to be used with, which is especially helpful in larger codebases or when using multiple modules.

Reply:
And why have `get_env_variable` in the first place: it is much more descriptive and hides the actual implementation for getting the env variable, which we are not interested in. It is immediately clear what that does making the code more readable.